CromwellRefdiskManifestCreator/src/main/java/org/broadinstitute/manifestcreator/CromwellRefdiskManifestCreatorApp.java

```diff
@@ -0,0 +1,194 @@
+
+import java.io.*;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Stream;
+import java.util.zip.CRC32C;
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.partitioningBy;
+
+public class CromwellRefdiskManifestCreatorApp {
+
+  private static final Logger logger = LogManager.getLogger(CromwellRefdiskManifestCreatorApp.class);
+
+  public static void main(String[] args) throws IOException, InterruptedException {
+    Configurator.setRootLevel(Level.INFO);
+
+    Arguments inputArguments = parseArguments(args);
+
+    ReferenceDiskManifest manifest = createManifestForDirectory(inputArguments);
+
+    logger.info("Writing file to disk...");
+    File manifestFile = new File(inputArguments.manifestFilePath);
+    if (manifestFile.exists() || !manifestFile.createNewFile()) {
+      logger.error("File {} already exists or cannot be created.", manifestFile.getAbsolutePath());
+      printUsageAndExit();
+    } else {
+      new ObjectMapper().writerWithDefaultPrettyPrinter().writeValue(manifestFile, manifest);
+      logger.info("Completed");
+    }
+  }
+
+  static ReferenceDiskManifest createManifestForDirectory(Arguments inputArguments) throws InterruptedException {
+    File rootDirectory = new File(inputArguments.directoryToScan);
+    if (!rootDirectory.exists() || !rootDirectory.isDirectory()) {
+      logger.error("Root directory {} doesn't exist.", inputArguments.directoryToScan);
+      printUsageAndExit();
+    }
+
+    logger.info("Populating list of files to process...");
+    Stack<File> allFilesInRootRecurStack = prepopulateStackWithFiles(rootDirectory);
+    logger.info("Need to process {} files", allFilesInRootRecurStack.size());
+
+    ReferenceDiskManifest manifest = new ReferenceDiskManifest();
+    manifest.setImageIdentifier(inputArguments.imageName);
+
+    CountDownLatch countDownLatch = new CountDownLatch(inputArguments.nThreads);
+    ExecutorService executorService = Executors.newFixedThreadPool(inputArguments.nThreads);
+
+    Path rootPath = rootDirectory.toPath();
+    List<ReferenceFile> referenceFiles = Collections.synchronizedList(new ArrayList<>());
+    for (int i = 0; i < inputArguments.nThreads; i++) {
+      executorService.submit(() -> {
+        try {
+          doWork(rootPath, referenceFiles, allFilesInRootRecurStack, countDownLatch);
+        } catch (CRC32CException e) {
+          logger.error("Error occurred. Shutting down.", e);
+          System.exit(1);
+        }
+      });
+    }
+
+    countDownLatch.await();
+    executorService.shutdown();
+
+    manifest.getFiles().addAll(referenceFiles);
+    logger.info("Finished created manifest. Total number of reference files: {}", manifest.getFiles().size());
+
+    return manifest;
+  }
+
+  private static void doWork(Path rootPath,
```

Comment:
I'd rename this something like `processFileStackIntoReferences`. Or just something a little more descriptive...